### PR TITLE
Change RunOnce duration plugin to act as a limit instead of override

### PR DESCRIPTION
**Pull request by csrwng**
Makes the following changes to the RunOnceDuration plugin:

- Intercepts only Creates instead of Create and Update
- Renames the config field to 'Limit' instead of 'Override'
- Prevents overwriting a set value on the pod if it is less than the configured limit

### CODE REVIEW

**Comment by csrwng** [comment]
@liggitt @smarterclayton ptal

File: pkg/quota/admission/runonceduration/api/types.go [hunk 1]
```diff
@@ -5,18 +5,18 @@ import (
...
 	// won't make any changes to the pod
-	ActiveDeadlineSecondsOverride *int64
+	ActiveDeadlineSecondsLimit *int64
 }
 
-// ActiveDeadlineSecondsOverrideAnnotation can be set on a project to override the number of
+// ActiveDeadlineSecondsLimitAnnotation can be set on a project to limit the number of
 // seconds that a run-once pod can be active in that project
-const ActiveDeadlineSecondsOverrideAnnotation = "openshift.io/active-deadline-seconds-override"
+const ActiveDeadlineSecondsLimitAnnotation = "openshift.io/active-deadline-seconds-limit"
```
Comment (liggitt) [inline pkg/quota/admission/runonceduration/api/types.go]: any doc need to be changed for this?

Comment (csrwng) [inline pkg/quota/admission/runonceduration/api/types.go]: Yes, this will require a follow-up doc PR

**Comment by csrwng** [comment]
@smarterclayton given that the existing config has been in since v1.1.2, is it acceptable to change the name of the config field?

**Comment by smarterclayton** [comment]
Generally no.

**Comment by csrwng** [comment]
So... would it be ok to still call it override, even though it's more like a max now?

**Comment by smarterclayton** [comment]
Add a TODO that we want to rename it in the future.  We can't rename
without updating the config version.

**Comment by csrwng** [comment]
Added a commit to keep existing external names. 
[test]

**Comment by csrwng** [comment]
[test]

File: pkg/quota/admission/runonceduration/admission.go [hunk 2]
```diff
@@ -114,14 +114,27 @@ func (a *runOnceDuration) applyProjectAnnotationOverride(namespace string, pod *
...
 	if err != nil {
-		return false, fmt.Errorf("cannot parse the ActiveDeadlineSeconds override (%s) for project %s: %v", override, ns.Name, err)
+		return false, fmt.Errorf("cannot parse the ActiveDeadlineSeconds limit (%s) for project %s: %v", limit, ns.Name, err)
 	}
-	pod.Spec.ActiveDeadlineSeconds = &overrideInt64
+	pod.Spec.ActiveDeadlineSeconds = minInt64(&limitInt64, pod.Spec.ActiveDeadlineSeconds)
 	return true, nil
 }
+
+func minInt64(a, b *int64) *int64 {
```
Comment (kargakis) [inline pkg/quota/admission/runonceduration/admission.go]: fyi, there is https://github.com/kubernetes/kubernetes/blob/master/pkg/util/integer/integer.go

**Comment by csrwng** [comment]
@kargakis thx, using the kube function now. 

@smarterclayton should this be considered for 3.2?

**Comment by smarterclayton** [comment]
What's the risk to online?  @bparees can you weigh in?

**Comment by csrwng** [comment]
This is more @abhgupta's area

**Comment by abhgupta** [comment]
@csrwng Online is fine with the current limit of 1 hour for runonce pods. If a user cannot set it any lower, that is fine for now. The only impacted use case is that users are trying to optimize their resource usage and want to restrict their runonce pod(s) to a smaller duration so that they can do early reclamation pod/cpu/memory quota for other purposes. I am not too concerned about this use case for Online on day 1 and if users really want to terminate their pods, they still have the ability to do so to reclaim their quota/resources.

**Comment by abhgupta** [comment]
@csrwng @smarterclayton Just realized that deployment cancellations are handled by explicitly setting the ActiveDeadlineSeconds on the deployer/hook pods to 0. Since this is prevented by the plugin, the functionality is broken and users won't be able to cancel a wedged deployment and be stuck with it for the full hour, preventing new deployments from being triggered.

I believe this is a blocker.

**Comment by smarterclayton** [comment]
Approved for 1.2

**Comment by csrwng** [comment]
@liggitt or @deads2k can you ptal?

File: pkg/quota/admission/runonceduration/admission.go [hunk 3]
```diff
@@ -114,14 +115,26 @@ func (a *runOnceDuration) applyProjectAnnotationOverride(namespace string, pod *
...
 	if err != nil {
-		return false, fmt.Errorf("cannot parse the ActiveDeadlineSeconds override (%s) for project %s: %v", override, ns.Name, err)
+		return false, fmt.Errorf("cannot parse the ActiveDeadlineSeconds limit (%s) for project %s: %v", limit, ns.Name, err)
 	}
-	pod.Spec.ActiveDeadlineSeconds = &overrideInt64
+	pod.Spec.ActiveDeadlineSeconds = Int64MinP(&limitInt64, pod.Spec.ActiveDeadlineSeconds)
 	return true, nil
 }
+
+func Int64MinP(a, b *int64) *int64 {
```
Comment (liggitt) [inline pkg/quota/admission/runonceduration/admission.go]: don't export this

File: pkg/quota/admission/runonceduration/api/types.go [hunk 4]
```diff
@@ -5,18 +5,19 @@ import (
...
 	// won't make any changes to the pod
-	ActiveDeadlineSecondsOverride *int64
+	ActiveDeadlineSecondsLimit *int64
 }
 
-// ActiveDeadlineSecondsOverrideAnnotation can be set on a project to override the number of
+// ActiveDeadlineSecondsLimitAnnotation can be set on a project to limit the number of
 // seconds that a run-once pod can be active in that project
-const ActiveDeadlineSecondsOverrideAnnotation = "openshift.io/active-deadline-seconds-override"
+// TODO: this label needs to change to reflect its function. It's a limit, not an override.
```
Comment (liggitt) [inline pkg/quota/admission/runonceduration/api/types.go]: note in the TODO that we are keeping it this way for compatibility, so someone doesn't just change it as cleanup without considering that

File: pkg/quota/admission/runonceduration/api/validation/validation.go [hunk 5]
```diff
@@ -9,11 +9,11 @@ import (
...
 func ValidateRunOnceDurationConfig(config *api.RunOnceDurationConfig) field.ErrorList {
 	allErrs := field.ErrorList{}
-	if config == nil || config.ActiveDeadlineSecondsOverride == nil {
+	if config == nil || config.ActiveDeadlineSecondsLimit == nil {
 		return allErrs
 	}
-	if *config.ActiveDeadlineSecondsOverride <= 0 {
-		allErrs = append(allErrs, field.Invalid(field.NewPath("activeDeadlineSecondsOverride"), config.ActiveDeadlineSecondsOverride, "must be greater than 0"))
+	if *config.ActiveDeadlineSecondsLimit <= 0 {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("activeDeadlineSecondsLimit"), config.ActiveDeadlineSecondsLimit, "must be greater than 0"))
```
Comment (liggitt) [inline pkg/quota/admission/runonceduration/api/validation/validation.go]: field name should still be activeDeadlineSecondsOverride

**Comment by liggitt** [comment]
just a couple comments, needs rebase, LGTM otherwise

**Comment by csrwng** [comment]
thx, made review changes

**Comment by csrwng** [comment]
[test]

**Comment by csrwng** [comment]
[test]

**Comment by csrwng** [comment]
[merge]

**Comment by xingxingxia** [comment]
@csrwng Works. But I must use activeDeadlineSeconds**Override** in master config file. If use activeDeadlineSeconds**Limit**, seems it is not recognized and the setting does not take effect.
You commented `Renames the config field to 'Limit' instead of 'Override'`. It is more straightforward to keep consistent name in master config file as well, isn't it? Thanks!

**Comment by csrwng** [comment]
@xingxingxia I only changed it to activeDeadlineSecondsLimit in the internal API but left it the same as before in the external API, which means you still need to use  activeDeadlineSecondsOverride in the master-config.yaml. Therefore, existing documentation should still be valid.

**Comment by xingxingxia** [comment]
OK, got it

**Pull request merged**